### ISSUE TITLE
Document passing a custom app wrapper

### ISSUE THREAD

**Issue by jukefr** [denoland/fresh#338]
I think fresh should allow passing a custom document wrapper in the same manner next does.
See https://github.com/denoland/fresh/blob/8d118a2d3055a7b5e734a9f2af838fdc14fe5c32/src/server/render.tsx#L270.
This would for example allow to set twind classes on the html element to support light/dark color schemes at a system level via media preferences, and also manually (see https://tailwindcss.com/docs/dark-mode#toggling-dark-mode-manually=)

It would probably need abstracting of this template function some more since it contains parts that you might want to hide from the user (i see some dangerouslySetInnerHTML etc)

To give a bit more of an idea about the context the first thing I tried after generating a project with twind is changing the darkMode to `media` and setting a dark mode background class property on the page, however to handle changing the background of the html element i would need to set the class on there or on the body at least. Currently the only way this would be possible is to make a container that has the same size as body which I find a bit "hacky" and not really intended practice from what I gather from tailwind docs.

It would probably allow for more flexibility outside of this specific use case also.

See https://nextjs.org/docs/advanced-features/custom-document for inspiration maybe.

**EDIT: this appears to actually be possible already although not documented (as far as I could tell) https://github.com/denoland/fresh/commit/04a32f34987fbd59ce646967f216b2cca072a58f**

**Comment by excessivecoding** [comment]
Facing the same issue with dark mode and also repeating element like `title`, etc.

I would love to have a more Remix approach where the DOM is completely exposed rather than the Next.js one with a `_app.jsx`, but this is just an opinion

**Comment by m4rocks** [comment]
+1 would be cool to see this. but it's already kind of possible by using a custom wrapper component for every route

**Comment by excessivecoding** [comment]
There is limitation on what you can do with custom wrapper though. For example changing property on the root `html` tag like `lang` and `dir` property

Edit: seems like it's possible to change `lang` just not `dir` and other html attributes

**Comment by EccentricVamp** [comment]
> Facing the same issue with dark mode and also repeating element like `title`, etc.

Same here. I was excited to see first class support for Twind, but Fresh doesn't seem to allow developers to easily set the CSS class of the `<html>` or `<body>` so no global styles for me.

> I would love to have a more Remix approach where the DOM is completely exposed

I agree. An exposed template also seems like a much more maintainable approach than adding fields to `TemplateOptions` for everything that could possibly be changed in the template.

Disclaimer: I come from a server-rendered ASP.NET Core background with no experience in JavaScript middleware frameworks. Not being able to directly modify the template (we call it _Layout.cshtml in ASP.NET Core) was really confusing to me. But I might just be biased.

**Comment by mah51** [comment]
I don't know if this is a recommended approach or not, however creating an _app.tsx within the routes/ directory the class attribute can be added to the html tag with the following code:

```
/** @jsx h */
import { h } from 'preact';
import { AppProps } from '$fresh/server.ts';

export default function App(props: AppProps) {
    return (
        <html class='dark'>
            <props.Component />
        </html>
    );
}
```

Hope this helps in the meantime!

**Comment by jukefr** [comment]
There does in fact appear to be an undocumented app wrapper feature available since https://github.com/denoland/fresh/commit/04a32f34987fbd59ce646967f216b2cca072a58f (did not pick this up at first quick glance at the source oops)
I guess it would be more adequate to switch this issue to adding documentation about it.

**Comment by EccentricVamp** [comment]
Relevant comment from a separate issue: https://github.com/denoland/fresh/issues/491#issuecomment-1181769820

TL;DR: `_app.tsx` is undocumented because it might be replaced

**Comment by hashrock** [comment]
Close with documenting `_app.tsx`. https://github.com/denoland/fresh/pull/494

Please reopen this issue if you think it doesn't meet your needs.
